functions/src/resolve-market.ts

```diff
@@ -31,140 +41,204 @@ export const resolveMarket = functions
+    probabilityInt:
+      contract.outcomeType == 'BINARY'
+        ? getProbability(contract) * 100
+        : undefined,
+    resolutions:
+      contract.outcomeType == 'FREE_RESPONSE'
+        ? getFreeResponseResolutions(contract)
+        : undefined,
+  }
+  contract.description = contract.description.concat(
+    `\n\n\nContract resolved automatically.`
+  )
+
+  return await privateResolveMarket(contract, data)
+}
+
+const getFreeResponseResolutions = (contract: Contract & FreeResponse) => {
+  const answersWithProbs = getAnswersWithProbs(contract)
+  const totalProb = sum(Object.values(answersWithProbs))
+  return mapValues(answersWithProbs, (prob) => (100 * prob) / totalProb)
+}
+
+const getAnswersWithProbs = (contract: Contract & FreeResponse) => {
+  const answers: { [id: string]: number } = {}
+  for (const answer of contract.answers) {
+    answers[answer.id] = getDpmOutcomeProbability(
+      contract.totalShares,
+      answer.id
+    )
+  }
+  return answers
+}
+
+const privateResolveMarket = async (
+  contract: Contract,
+  data: {
+    outcome: resolution
+    value?: number
+    probabilityInt?: number
+    resolutions?: { [outcome: string]: number }
+  }
+) => {
+  const { creatorId, id, outcomeType, closeTime } = contract
+  const { outcome, probabilityInt, resolutions, value } = data
+  switch (outcomeType) {
+    case 'FREE_RESPONSE':
+      if (
+        isNaN(+outcome) &&
+        !(outcome === 'MKT' && resolutions) &&
+        outcome !== 'CANCEL'
+      )
+        return { status: 'error', message: 'Invalid outcome' }
+      break
+    case 'NUMERIC':
+      if (isNaN(+outcome) && outcome !== 'CANCEL')
+        return { status: 'error', message: 'Invalid outcome' }
+      break
+    case 'BINARY':
+      if (!RESOLUTIONS.includes(outcome))
+        return { status: 'error', message: 'Invalid outcome' }
       if (
-        outcomeType === 'BINARY' &&
-        probabilityInt !== undefined &&
+        probabilityInt &&
         (probabilityInt < 0 ||
           probabilityInt > 100 ||
           !isFinite(probabilityInt))
       )
         return { status: 'error', message: 'Invalid probability' }
+  }
 
-      if (creatorId !== userId)
-        return { status: 'error', message: 'User not creator of contract' }
+  if (value && !isFinite(value))
+    return { status: 'error', message: 'Invalid value' }
 
-      if (contract.resolution)
-        return { status: 'error', message: 'Contract already resolved' }
+  if (contract.resolution)
+    return { status: 'error', message: 'Contract already resolved' }
 
-      const creator = await getUser(creatorId)
-      if (!creator) return { status: 'error', message: 'Creator not found' }
+  const creator = await getUser(creatorId)
+  if (!creator) return { status: 'error', message: 'Creator not found' }
 
-      const resolutionProbability =
-        probabilityInt !== undefined ? probabilityInt / 100 : undefined
+  const resolutionProbability = probabilityInt
```

Comment:
This is incorrect if `probabilityInt` is 0. Switch back to checking `probabilityInt !== undefined`